Plugin service tests

charliescheer opened the pull request:
### Description

This PR adds unit tests to the parts of  WordPressKit-iOS that handle plugins.  The updates cover these files:
PluginDirectoryFeedPage.swift
PluginState.swift
SitePluginCapabilities.swift
SitePlugin.swift
PluginDirectoryEntry.swift
PluginServiceRemote.swift
PluginDirectoryServiceRemote.swift


### Testing Details
PR only deals with the addition of new tests, to test it, simply run the testing suite WordPressKitTests.  All tests should pass.

charliescheer committed:
Add tests to confirm equatable works for Plugin Page Directories

charliescheer committed:
Add mock data for plugin service remote featured plugins

charliescheer committed:
added unit test for Plugin service remote Get featured plugins succeeds

charliescheer committed:
updated testPluginFeedPageDirectoryEquatable to use different arsed responses for checking equatable

charliescheer committed:
added new file for PluginStateTests

charliescheer committed:
added mocks for plugin state for Jetpack and contact form 7

charliescheer committed:
created setup for PluginStateTests

charliescheer committed:
added tests for plugin remote service - plugin updates

charliescheer committed:
added tests for plugin service remote plugin update

charliescheer committed:
Added tests for PluginState functions

charliescheer committed:
added Plugin Capatabilities equatable test to pluginStateTests

charliescheer committed:
added tests to PluginServiceRemote to test getting featured plugins failing

charliescheer committed:
correcting incorrect podfile version

charliescheer committed:
added another test for checking the error catching fir fetching featured plugins in PluginServiceRemote

charliescheer committed:
Added tests for getting site plugins failing.  And added a test for the plugin service remote site id encode method

charliescheer committed:
added tests for plugin update failing

charliescheer committed:
added more json response mock data for the plugin service remote tests

charliescheer committed:
added function to test plugin install

charliescheer committed:
refactored creating post request stubs to a single helper test method

charliescheer committed:
added functions for testing plugin modify and remove failing with malformed json

charliescheer committed:
Refactored the setup for the plugin state tests to not rely on creating a service remote each time, and instead worked with a base set of PluginState objects

charliescheer committed:
Added tests to check SitePluginCapabiliies equatable works

charliescheer committed:
Added a function to test the output of PluginDirectoryEntry.descriptionText

charliescheer committed:
added extension to PluginDirectoryEntry to make the local functions to the file internal static functions for ease of testing

charliescheer committed:
Added test to check the trim change log method

charliescheer committed:
removed previous extentsion and internau static functions, found an easier way to do the same thing

charliescheer committed:
added function to test codable for plugin directory entry

charliescheer committed:
added function to test codable for pluginstate

charliescheer committed:
Created base for AtomicTransferTests to test the AtomicTransferServiceRemote

charliescheer committed:
Revert "Created base for AtomicTransferTests to test the AtomicTransferServiceRemote"

This reverts commit 482104a4de6dc9065f446c803910b5adce09c7e3.

charliescheer committed:
Revert "added function to test codable for pluginstate"

This reverts commit 808d7e4491efc86337eaebe9f8f099a3c1a67718.

charliescheer committed:
Added a new test to check that the PluginDirectoryFeedPage Decoder works

charliescheer committed:
changed PluginStateTests to use locally defined objects instead of decoded JSON objects

charliescheer committed:
Added tests for encoding and decogind PluginState.UpdateState

charliescheer committed:
removed custom set up and tear down for PluginStateTests

charliescheer committed:
removed class properties for pluginstatetests

charliescheer committed:
Changed SitePluginTests to use locally defined variables instead of creating them from JSON data

charliescheer committed:
Removed Remote and REST test protocols from PluginStateTests and SitePluginsTests as they no longer needed the methods to stub responses from the WP API

charliescheer committed:
Removed the remote test cases from pluginstate tests and siteplugintests

charliescheer committed:
Added local files to do model tests against and updated tests to use it

charliescheer committed:
Added test for PluginDirectory encode functions

charliescheer committed:
Added tests to check equatable and slug output for PluginDirectoryFeeds

charliescheer committed:
Refactored PluginStateTests to use a mock object instead of local data

charliescheer committed:
Added helper methods to MockPluginStateProvider

charliescheer committed:
Undoing a previous change to PluginDirectoryEntry to not use static func

charliescheer committed:
Removed some faulty asserts from func testExtractHTMLTextOutput

charliescheer committed:
Moved model data from directory tests to MockPluginDirectoryEntryProvide

charliescheer committed:
Added tests to cover some additional function output cases

charliescheer committed:
Added additional modify plugin function failure tests

charliescheer committed:
Added tests to validate PluginDirectoryFeedEndpoint responses

charliescheer committed:
Added dynamic mock protocol

charliescheer committed:
Added PluginState tests with dynamically generated PluginStates

charliescheer committed:
Added test for plugin Service remote decoding PluginState dynamic JSON

jkmassel commented on WordPressKit/PluginDirectoryEntry.swift:
```diff
@@ -12,10 +12,10 @@ public struct PluginDirectoryEntry {
-    private let installationHTML: String?
-    private let faqHTML: String?
-    private let changelogHTML: String?
+    internal let descriptionHTML: String?
```
`internal` should be the default, so it's not strictly necessary to have them here, but it doesn't hurt anything either :)

jkmassel commented on WordPressKit/PluginServiceRemote.swift:
```diff
@@ -16,7 +16,7 @@ public class PluginServiceRemote: ServiceRemoteWordPressComREST {
                 return
             }
-
+            
```
Might need to set Xcode to remove empty whitespace in lines

charliescheer replied:
Done, didn't know that was an option :)
Thanks!

jkmassel commented on WordPressKitTests/Mock Data/plugin-activate-malformed-response.json:
```diff
@@ -0,0 +1,18 @@
+{
+    "slug" "code-snippets",
```
This response could probably be just this one line if you're just testing for invalid JSON

jkmassel replied:
Applies to others as well!

jkmassel commented on WordPressKitTests/MockPluginDirectoryEntryProvider.swift:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    static func getJetpackDescriptionHTML() -> String {
+        let description = "<p>Keep any WordPress site secure, increase traffic, and engage your readers.</p>\n<h4>Traffic and SEO Tools</h4>\n<p>Traffic is the lifeblood of any website. Jetpack includes:</p>\n<ul>\n<li>[free] Site stats and analytics</li>\n<li>[free] Automatic sharing on Facebook, Twitter, LinkedIn, Tumblr, Reddit, and WhatsApp</li>\n<li>[free] Related posts</li>\n<li>[paid] Sear ...
```
WDYT about putting this in a JSON file and parsing it in? It's a lot of boilerplate IMHO? Not necessarily wrong...tbh I'm not sure if that's best either but something I was thinking about.

jkmassel commented on WordPressKitTests/MockPluginStateProvider.swift:
```diff
@@ -0,0 +1,64 @@
+        var data = Data()
+        let encoder = JSONEncoder()
+        
+        do {
```
Because this is a test helper, it might make sense to just `throw` out of this method instead of handling it – if it's not working, that failure should probably bubble up to the individual test.

jkmassel commented on WordPressKitTests/PluginDirectoryTests.swift:
```diff
@@ -77,6 +77,28 @@ class PluginDirectoryTests: XCTestCase {
         }
     }
+    
+    func testValidatePluginDirectoryFeedResponseSucceeds() {
```
These tests can `throw` on their own, and if they do it'll be counted as a failure – no need to catch it! (this is a relatively new feature that may require a newer Xcode version – if so, we can update it in CI)

jkmassel commented on WordPressKitTests/PluginDirectoryTests.swift:
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
+    }
+    
+    func testPluginFeedPageDirectoryEquatable() {
+        let jetpackMockPath = Bundle(for: type(of: self)).path(forResource: "plugin-directory-jetpack", ofType: "json")!
```
We repeat this three-line pattern a lot – perhaps it could be extracted into a helper?

jkmassel commented on WordPressKitTests/PluginDirectoryTests.swift:
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
+        do {
+            let directoryEntry = try PluginDirectoryEntry(responseObject: response)
+
+            XCTAssertEqual(directoryEntry.name, "Jetpack by WordPress.com")
```
This is good, but it comes back to what I mentioned about dynamic naming – if you can get this dynamically generated (ideally using non-ascii characters, HTML entities, whatever other edge cases you can think of), it'll result in a better test because it stressed the code rather than just validating inputs to outputs.

jkmassel commented on WordPressKitTests/PluginDirectoryTests.swift:
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
+    }
+
+    func testPluginDirectoryFeedTypeSlugReturn() {
+        let pluginDirectoryFeedTypeNewest = PluginDirectoryFeedType.newest
+        let pluginDirectoryFeedTypePopular = PluginDirectoryFeedType.popular
+        let pluginDirectoryFeedTypeSearch = PluginDirectoryFeedType.search(term: "blocks")
```
Nitpick – these lines may not be necessary – you could probably use `PluginDirectoryFeedType.newest.slug` in the assertion?

jkmassel commented on WordPressKitTests/PluginServiceRemoteTests.swift:
```diff
@@ -65,4 +84,438 @@ class PluginServiceRemoteTests: RemoteTestCase, RESTTestable {
+        waitForExpectations(timeout: timeout, handler: nil)
+    }
+    
+    func testPluginIDEncoding() {
```
This is a great type of test to have

jkmassel reviewed (approved):
This looks very good – I've left a few little comments, but overall I think this is a good set of tests to validate this area of the project!

jkmassel commented on WordPressKitTests/PluginDirectoryTests.swift:
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
+        }
+    }
+    
+    func testPluginFeedPageDirectoryEquatable() {
```
One other thing I'd suggest adding tests for – are there any feed page directories that superficially might look the same, but actually aren't? Is that something that can be tested?

scheercharlie committed:
Removed declaration of Internal in Plugins where it would be the default

scheercharlie committed:
Updated invalid plugin JSONmocks to contain simpler examples of bad json

scheercharlie committed:
Changed how MockPluginProvider fetched encoded update states

Changed where the encode errors were handled so that they would handled in the test instead of in the provider it's self

scheercharlie committed:
Simplified  throw tests to utilize XCTTests check for throwing functions

scheercharlie committed:
Refactored getting JSON data for tests into a mock provider

scheercharlie committed:
Simplified the test inputs for testPluginDirectoryFeedTypeSlugReturn

scheercharlie committed:
Added a test to see if similar PluginDirectories would be Equatable

scheercharlie committed:
Updated MockPluginStateProvider to be able to provide dynamic json

scheercharlie committed:
Removed unused variable name declarations to hide some swift warnings

scheercharlie committed:
Moved MockPluginDirectoryEntry HTML descriptions to JSON mock files
